src/coreclr/tools/Common/Internal/Runtime/DehydratedData.cs

```diff
@@ -0,0 +1,116 @@
+    /// </summary>
+    /// <remarks>
+    /// The instructions use a variable length encoding and are split in two parts:
+    /// the instruction command kind and command data (payload).
+    /// The payload is an integer. Instruction kind and payload can fit into a single
+    /// byte, the encoding is one byte. Bigger payloads produce bigger instructions.
+    /// </remarks>
+    internal static class DehydratedDataCommand
+    {
+        public const byte Copy = 0x00;
+        public const byte ZeroFill = 0x01;
+        public const byte RelPtr32Reloc = 0x02;
+        public const byte PtrReloc = 0x03;
+
+        private const byte DehydratedDataCommandMask = 0x03;
+        private const int DehydratedDataCommandPayloadShift = 2;
+
+        private const int MaxRawShortPayload = (1 << (8 - DehydratedDataCommandPayloadShift)) - 1;
+        private const int MaxExtraPayloadBytes = 3;
+        private const int MaxShortPayload = MaxRawShortPayload - MaxExtraPayloadBytes;
+
+        public static byte EncodeShort(int command, int commandData)
+        {
+            Debug.Assert((command & DehydratedDataCommandMask) == command);
+            Debug.Assert(commandData <= MaxShortPayload);
+            return (byte)(command | (commandData << DehydratedDataCommandPayloadShift));
+        }
+
+        public static int Encode(int command, int commandData, byte[] buffer)
+        {
+            Debug.Assert((command & DehydratedDataCommandMask) == command);
+            int remainingData = commandData - MaxShortPayload;
+            if (remainingData <= 0)
+            {
+                buffer[0] = EncodeShort(command, commandData);
+                return 1;
+            }
+
+            int numExtraBytes = 0;
+            for (; remainingData != 0; remainingData >>= 8)
+                buffer[++numExtraBytes] = (byte)remainingData;
+            Debug.Assert(numExtraBytes <= MaxExtraPayloadBytes, "Decoder assumes this");
+
+            buffer[0] = (byte)(command | ((MaxShortPayload + numExtraBytes) << DehydratedDataCommandPayloadShift));
+            return 1 + numExtraBytes;
+        }
+
+        public static unsafe byte* Decode(byte* pB, out int command, out int payload)
+        {
+            byte b = *pB;
+            command = b & DehydratedDataCommandMask;
+            payload = b >> DehydratedDataCommandPayloadShift;
+            int extraBytes = payload - MaxShortPayload;
+            if (extraBytes > 0)
+            {
+                payload = *++pB;
+                if (extraBytes > 1)
+                {
+                    payload += *++pB << 8;
+                    if (extraBytes > 2)
+                        payload += *++pB << 16;
+                }
+
+                payload += MaxShortPayload;
+            }
+
+            return pB + 1;
+        }
+
+#if false
+        static void Main()
+        {
+            int command, payload;
+
+            byte[] buf = new byte[5];
+            Debug.Assert(Encode(1, 0, buf) == 1);
```

Comment:
This is under `#if false` - it was my debugging helper to make sure I didn't mess up the bit packing (this file can be compiled on its own with the region enabled).

I'm still debating whether to check it in, delete it, or spend more time converting it to a unit test that is never going to fail because nobody will want to touch this anyway.